File: runtime/testdata/node-runtime/src/push.ts
import fs from 'node:fs';
import nodePath from 'node:path';
import webpush from 'web-push';
import type { PushSubscription } from 'web-push';
import { PRETTYD_STATE_DIR } from './config.js';

const VAPID_PATH = nodePath.join(PRETTYD_STATE_DIR, 'vapid.json');
const SUBSCRIPTIONS_PATH = nodePath.join(PRETTYD_STATE_DIR, 'push-subscriptions.json');
const VAPID_SUBJECT = 'mailto:pretty-pty@localhost';

interface VapidKeys {
  publicKey: string;
  privateKey: string;
}

export type PushSubscriptionRecord = PushSubscription;

export interface PushPayload {
  title: string;
  body?: string;
  data?: Record<string, unknown>;
}

function ensureStateDir(): void {
  fs.mkdirSync(PRETTYD_STATE_DIR, { recursive: true, mode: 0o700 });
}

function isObject(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null;
}

function isVapidKeys(value: unknown): value is VapidKeys {
  return isObject(value)
    && typeof value.publicKey === 'string'
    && value.publicKey.length > 0
    && typeof value.privateKey === 'string'
    && value.privateKey.length > 0;
}

function isPushSubscription(value: unknown): value is PushSubscriptionRecord {
  if (!isObject(value)) return false;
  const keys = value.keys;
  return typeof value.endpoint === 'string'
    && value.endpoint.length > 0
    && (value.expirationTime === undefined || value.expirationTime === null || typeof value.expirationTime === 'number')
    && isObject(keys)
    && typeof keys.p256dh === 'string'
    && keys.p256dh.length > 0
    && typeof keys.auth === 'string'
    && keys.auth.length > 0;
}

function readVapidKeys(): VapidKeys | null {
  try {
    const parsed = JSON.parse(fs.readFileSync(VAPID_PATH, 'utf8')) as unknown;
    return isVapidKeys(parsed) ? parsed : null;
  } catch {
    return null;
  }
}

function writeVapidKeys(keys: VapidKeys): void {
  ensureStateDir();
  fs.writeFileSync(VAPID_PATH, JSON.stringify(keys, null, 2), { mode: 0o600 });
}

function getVapidKeys(): VapidKeys {
  const existing = readVapidKeys();
  if (existing) return existing;
  const generated = webpush.generateVAPIDKeys();
  writeVapidKeys(generated);
  return generated;
}

function loadSubscriptions(): PushSubscriptionRecord[] {
  try {
    const parsed = JSON.parse(fs.readFileSync(SUBSCRIPTIONS_PATH, 'utf8')) as unknown;
    if (!Array.isArray(parsed)) return [];
    return parsed.filter(isPushSubscription);
  } catch {
    return [];
  }
}

function saveSubscriptions(subscriptions: PushSubscriptionRecord[]): void {
  ensureStateDir();
  fs.writeFileSync(SUBSCRIPTIONS_PATH, JSON.stringify(subscriptions, null, 2), { mode: 0o600 });
}

function getPushStatusCode(err: unknown): number | null {
  if (err instanceof webpush.WebPushError) return err.statusCode;
  if (!isObject(err)) return null;
  const statusCode = err.statusCode;
  return typeof statusCode === 'number' ? statusCode : null;
}

export function getVapidPublicKey(): string {
  return getVapidKeys().publicKey;
}

export function addSubscription(subscription: unknown): void {
  if (!isPushSubscription(subscription)) {
    throw new Error('invalid push subscription');
  }
  const subscriptions = loadSubscriptions();
  const withoutExisting = subscriptions.filter((s) => s.endpoint !== subscription.endpoint);
  saveSubscriptions([...withoutExisting, subscription]);
}

export function removeSubscription(endpoint: string): void {
  const subscriptions = loadSubscriptions();
  saveSubscriptions(subscriptions.filter((s) => s.endpoint !== endpoint));
}

export async function sendPush(payload: PushPayload): Promise<void> {
  let keys: VapidKeys;
  let subscriptions: PushSubscriptionRecord[];
  try {
    keys = getVapidKeys();
    subscriptions = loadSubscriptions();
  } catch {
    return;
  }

  for (const subscription of subscriptions) {
    try {
      await webpush.sendNotification(subscription, JSON.stringify(payload), {
        vapidDetails: {
          subject: VAPID_SUBJECT,
          publicKey: keys.publicKey,
          privateKey: keys.privateKey
        },
        TTL: 60 * 60,
        urgency: 'normal'
      });
    } catch (err) {
      const statusCode = getPushStatusCode(err);
      if (statusCode === 404 || statusCode === 410) {
        try { removeSubscription(subscription.endpoint); } catch { /* best effort */ }
      }
    }
  }
}

File: runtime/internal/session/push.go
package session

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"os"
	"path/filepath"
	"sync"

	webpush "github.com/SherClockHolmes/webpush-go"
)

const vapidSubject = "mailto:sessions@localhost"

type VAPIDKeys struct {
	PublicKey  string `json:"publicKey"`
	PrivateKey string `json:"privateKey"`
}

type PushSubscription struct {
	Endpoint       string          `json:"endpoint"`
	ExpirationTime json.RawMessage `json:"expirationTime,omitempty"`
	Keys           struct {
		P256dh string `json:"p256dh"`
		Auth   string `json:"auth"`
	} `json:"keys"`
}

type PushPayload struct {
	Title string         `json:"title"`
	Body  string         `json:"body,omitempty"`
	Data  map[string]any `json:"data,omitempty"`
}

type PushService struct {
	root              string
	vapidPath         string
	subscriptionsPath string
	mu                sync.Mutex
	httpClient        webpush.HTTPClient
}

func NewPushService(stateRoot string) *PushService {
	return &PushService{
		root:              stateRoot,
		vapidPath:         filepath.Join(stateRoot, "vapid.json"),
		subscriptionsPath: filepath.Join(stateRoot, "push-subscriptions.json"),
	}
}

func (p *PushService) SetHTTPClient(client webpush.HTTPClient) { p.httpClient = client }

func (p *PushService) HasSubscriptions() bool {
	p.mu.Lock()
	defer p.mu.Unlock()
	return len(p.loadSubscriptionsLocked()) > 0
}

func (p *PushService) VAPIDPublicKey() (string, error) {
	p.mu.Lock()
	defer p.mu.Unlock()
	keys, err := p.vapidKeysLocked()
	return keys.PublicKey, err
}

func (p *PushService) AddSubscription(value any) error {
	raw, err := json.Marshal(value)
	if err != nil {
		return errors.New("invalid push subscription")
	}
	var subscription PushSubscription
	if json.Unmarshal(raw, &subscription) != nil || !validSubscription(subscription) {
		return errors.New("invalid push subscription")
	}
	p.mu.Lock()
	defer p.mu.Unlock()
	subscriptions := p.loadSubscriptionsLocked()
	filtered := subscriptions[:0]
	for _, existing := range subscriptions {
		if existing.Endpoint != subscription.Endpoint {
			filtered = append(filtered, existing)
		}
	}
	return p.saveSubscriptionsLocked(append(filtered, subscription))
}

func (p *PushService) RemoveSubscription(endpoint string) error {
	p.mu.Lock()
	defer p.mu.Unlock()
	subscriptions := p.loadSubscriptionsLocked()
	filtered := subscriptions[:0]
	for _, subscription := range subscriptions {
		if subscription.Endpoint != endpoint {
			filtered = append(filtered, subscription)
		}
	}
	return p.saveSubscriptionsLocked(filtered)
}

func (p *PushService) Send(ctx context.Context, payload PushPayload) {
	p.mu.Lock()
	subscriptions := p.loadSubscriptionsLocked()
	if len(subscriptions) == 0 {
		p.mu.Unlock()
		return
	}
	keys, err := p.vapidKeysLocked()
	p.mu.Unlock()
	if err != nil {
		return
	}
	message, err := json.Marshal(payload)
	if err != nil {
		return
	}
	for _, subscription := range subscriptions {
		response, sendErr := webpush.SendNotificationWithContext(ctx, message, &webpush.Subscription{
			Endpoint: subscription.Endpoint,
			Keys:     webpush.Keys{P256dh: subscription.Keys.P256dh, Auth: subscription.Keys.Auth},
		}, &webpush.Options{
			HTTPClient: p.httpClient, Subscriber: vapidSubject,
			VAPIDPublicKey: keys.PublicKey, VAPIDPrivateKey: keys.PrivateKey,
			TTL: 60 * 60, Urgency: webpush.UrgencyNormal,
		})
		status := 0
		if response != nil {
			status = response.StatusCode
			_ = response.Body.Close()
		}
		if (sendErr != nil && (status == http.StatusNotFound || status == http.StatusGone)) ||
			status == http.StatusNotFound || status == http.StatusGone {
			_ = p.RemoveSubscription(subscription.Endpoint)
		}
	}
}

func (p *PushService) vapidKeysLocked() (VAPIDKeys, error) {
	if encoded, err := os.ReadFile(p.vapidPath); err == nil {
		var keys VAPIDKeys
		if json.Unmarshal(encoded, &keys) == nil && keys.PublicKey != "" && keys.PrivateKey != "" {
			return keys, nil
		}
	}
	privateKey, publicKey, err := webpush.GenerateVAPIDKeys()
	if err != nil {
		return VAPIDKeys{}, fmt.Errorf("generate VAPID keys: %w", err)
	}
	keys := VAPIDKeys{PublicKey: publicKey, PrivateKey: privateKey}
	if err := p.writeJSONLocked(p.vapidPath, keys); err != nil {
		return VAPIDKeys{}, err
	}
	return keys, nil
}

func (p *PushService) loadSubscriptionsLocked() []PushSubscription {
	encoded, err := os.ReadFile(p.subscriptionsPath)
	if err != nil {
		return []PushSubscription{}
	}
	var items []json.RawMessage
	if json.Unmarshal(encoded, &items) != nil {
		return []PushSubscription{}
	}
	subscriptions := make([]PushSubscription, 0, len(items))
	for _, item := range items {
		var subscription PushSubscription
		if json.Unmarshal(item, &subscription) == nil && validSubscription(subscription) {
			subscriptions = append(subscriptions, subscription)
		}
	}
	return subscriptions
}

func validSubscription(subscription PushSubscription) bool {
	if subscription.Endpoint == "" || subscription.Keys.P256dh == "" || subscription.Keys.Auth == "" {
		return false
	}
	if len(subscription.ExpirationTime) == 0 || string(subscription.ExpirationTime) == "null" {
		return true
	}
	var number json.Number
	return json.Unmarshal(subscription.ExpirationTime, &number) == nil
}

func (p *PushService) saveSubscriptionsLocked(subscriptions []PushSubscription) error {
	return p.writeJSONLocked(p.subscriptionsPath, subscriptions)
}

func (p *PushService) writeJSONLocked(path string, value any) error {
	if err := os.MkdirAll(p.root, 0o700); err != nil {
		return err
	}
	encoded, err := json.MarshalIndent(value, "", "  ")
	if err != nil {
		return err
	}
	if err := os.WriteFile(path, encoded, 0o600); err != nil {
		return err
	}
	return os.Chmod(path, 0o600)
}
